isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java

```diff
@@ -0,0 +1,126 @@
+      RelOptCluster relOptCluster,
+      CalciteCatalogReader catalogReader) {
+    this.relOptCluster = relOptCluster;
+    this.catalogReader = catalogReader;
+    this.extensions = extensions;
+    this.scalarFunctionConverter =
+        new ScalarFunctionConverter(
+            this.extensions.scalarFunctions(), relOptCluster.getTypeFactory());
+  }
+
+  public static RelNode convert(
+      Rel relRoot, RelOptCluster relOptCluster, CalciteCatalogReader calciteCatalogReader) {
+    return relRoot.accept(
+        new SubstraitRelNodeConverter(EXTENSION_COLLECTION, relOptCluster, calciteCatalogReader));
+  }
+
+  @Override
+  public RelNode visit(Aggregate aggregate) throws RuntimeException {
+    throw new UnsupportedOperationException("Aggregate is not supported in SubstraitToSql!");
+  }
+
+  @Override
+  public RelNode visit(EmptyScan emptyScan) throws RuntimeException {
+    throw new UnsupportedOperationException("EmptyScan is not supported in SubstraitToSql!");
+  }
+
+  @Override
+  public RelNode visit(Fetch fetch) throws RuntimeException {
+    throw new UnsupportedOperationException("Fetch is not supported in SubstraitToSql!");
+  }
+
+  @Override
+  public RelNode visit(Filter filter) throws RuntimeException {
+    RelNode input = filter.getInput().accept(this);
+    RexNode filterCondition =
+        filter
+            .getCondition()
+            .accept(
+                new ExpressionRexConverter(
+                    relOptCluster.getTypeFactory(), scalarFunctionConverter));
+    return LogicalFilter.create(input, filterCondition);
+  }
+
+  @Override
+  public RelNode visit(Join join) throws RuntimeException {
+    throw new UnsupportedOperationException("Join is not supported in SubstraitToSql!");
+  }
+
+  @Override
+  public RelNode visit(NamedScan namedScan) throws RuntimeException {
+    RelDataType rowType =
+        TypeConverter.convert(
+            relOptCluster.getTypeFactory(),
+            namedScan.getRecordType(),
+            namedScan.getInitialSchema().names());
+    CalciteSchema.TableEntry entry =
+        catalogReader.getRootSchema().getTable(namedScan.getNames().get(0), false);
+    if (entry == null) {
+      throw new RuntimeException("Table " + namedScan.getNames() + " not found in catalog");
+    }
+    RelOptTable table = RelOptTableImpl.create(catalogReader, rowType, entry, null /*rowcount*/);
+    return LogicalTableScan.create(relOptCluster, table, ImmutableList.of() /*hints*/);
+  }
+
+  @Override
+  public RelNode visit(Project project) throws RuntimeException {
+    RelNode child = project.getInput().accept(this);
+    List<RexNode> rexList =
+        project.getExpressions().stream()
+            .map(
+                expr ->
+                    expr.accept(
+                        new ExpressionRexConverter(
+                            relOptCluster.getTypeFactory(), scalarFunctionConverter)))
+            .toList();
+
+    return LogicalProject.create(
```

Comment:
I think we should probably be using a RelBuilder rathern invoking these static constructors individually. Or do you think that doesn't make sense?

Reply:
The current code only supports scan/project/filter, and the static constructors are relatively simple. 

I switched to RelBuilder to construct RelNode. However, I have to remove the checking of [1].  The RelBuilder seems to create a catalogreader and a new type system, different from the one when tables are parsed/added. 


1. https://github.com/substrait-io/substrait-java/blob/main/isthmus/src/main/java/io/substrait/isthmus/SqlToSubstrait.java#L255-L256